plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange.hs

```diff
@@ -126,6 +150,34 @@ findPosition pos root = go Nothing root
             startOfRight <- _start . _codeRange_range <$> V.headM right
             if pos < startOfRight then binarySearchPos left else binarySearchPos right
 
+-- | Traverses through the code range and children to a folding ranges
+findFoldingRanges :: CodeRange -> [FoldingRange]
+findFoldingRanges r@(CodeRange _ children _) =
+  let frRoot :: [FoldingRange] = case createFoldingRange r of
+        Just x  -> [x]
+        Nothing -> []
+
+      frChildren :: [FoldingRange] = concat $ V.toList $ fmap findFoldingRanges children
+   in frRoot ++ frChildren
+
+-- | Parses code range to folding range
+createFoldingRange :: CodeRange -> Maybe FoldingRange
+createFoldingRange (CodeRange (Range (Position lineStart charStart) (Position lineEnd charEnd)) _ ck) = do
+    let frk = crkToFrk ck
+    if lineStart == lineEnd
```

Comment:
why? this sort of thing _really_ needs a comment, ideally with an example showing why it's problematic. You spent lots of time figuring this out, and all that knowledge is going to be lost unless you write it down!

Reply:
PTAL, if the explanation in the comments is adequate or not?

I wrote a nice explanation in the remove duplicate as well as the function where we parse the root and its children.

Reply:
Good explanation. Are we 100% sure this is true, though? I feel like I've seen 1-line folding ranges in emacs where it turns
```
foo = ...
```
into
```
foo ...
```
which is a bit weird but not obviously wrong. Maybe we should pass them through and just let the client decide whether or not to use them? You said vscode ignores them already.

Reply:
Do people use this kind of foldings?

I guess we should let it be like this, there are a lot of foldings already and it might make things chaotic.

Also talked about this with @kokobd earlier

Reply:
Then maybe we should preserve the longest folding range on each line?

Reply:
> Then maybe we should preserve the longest folding range on each line?

I guess we should do this

Reply:
What happens if we send all the ranges to the client? I guess you tried it and it was bad? Worth writing it down so we know why we're doing it.

Reply:
Nothing happens, vscode handles it well.

But the code is difficult to test (reference: https://github.com/haskell/haskell-language-server/pull/3058#issuecomment-1242659776 and the commit above this)

And if there is a client that does not handle this, it might go bad, so handling it on hls itself

Reply:
I think if the client handles it then it's better to just give the client all the information! We can change it later if it's a problem.

I agree it's more effort to verify that the test output is correct, but that's useful work: we're going to return lots of small folding ranges, we _should_ think carefully about whether they're all right!

Reply:
Should I keep the code there and commented or should i remove it for handling the extra folding ranges?

Reply:
Then I vote to remove the deduplicating code if that doesn't change the user experience. You should leave a comment explaining the situation.

Reply:
Yeah, I think we can just a) not remove folding ranges on a single line and b) not remove folding ranges contained in other folding ranges, and add a note saying "we think the client can handle this, if not we could change to remove these in future".

Reply:
Okay, I’ll remove both of the filters and add a comment explaining the situation and edit the tests.

Reply:
Resolved